feat: ledger recovery

brianp opened the pull request:
Description
---
Utilize the view key from the ledger to perform recovery on a wallet, and support one-sided transactions.

Motivation and Context
---
Allow recovery.

How Has This Been Tested?
---
Manually

- [x] Software -> Software
    - [x] One-sided
    - [x] Interactive
- [x] Ledger -> Software
    - [x] One-sided
    - [x] Interactive
- [x] Ledger -> Ledger
    - [x] Interactive
    - [x] One-sided

- [x] Software recovery
- [x] Ledger recovery

What process can a PR reviewer use to test or verify this change?
---
Please run a manual test between wallets.

Breaking Changes
---

- [ ] None
- [x] Requires data directory on base node to be deleted
- [x] Requires hard fork
- [ ] Other - Please specify

brianp committed:
Add ledger function for providing the private view key

brianp committed:
Extend the software wallet to a struct

brianp committed:
Use the same branch key for comms secret key & view key

brianp committed:
Use the local view key for ledger

brianp committed:
Remove identity file code. The wallet only stores in db

brianp committed:
Ensure the public key is found when needed

brianp committed:
little refactor

brianp committed:
Update recovery prompts

brianp committed:
Copy fixes for consistency

brianp committed:
Make Alpha keys in software wallets derivable

brianp committed:
Make clippy happy

SWvheerden commented on base_layer/common_types/src/wallet_types.rs:
```diff
@@ -26,47 +26,38 @@ use std::{
 #[derive(Debug, Clone, Serialize, Deserialize)]
 pub struct LedgerWallet {
     account: u64,
     pub public_alpha: Option<RistrettoPublicKey>,
```
this should be of type `crate::types::publickey`

SWvheerden commented on base_layer/wallet/src/wallet.rs:
```diff
@@ -814,7 +819,7 @@ pub fn read_or_create_wallet_type<T: WalletBackend + 'static>(
     let comms_key_manager = KeyManager::<PublicKey, KeyDigest>::from(
         master_seed.clone(),
-        KEY_MANAGER_COMMS_SECRET_KEY_BRANCH_KEY.to_string(),
+        TransactionKeyManagerBranch::DataEncryption.get_branch_key(),
```
this does not seem correct

brianp replied:
Leaving a comment for any interested spectators.

*narrator* : It was in fact, not correct.

brianp committed:
Undo view key == comms key mixup

brianp committed:
Use PublicKey type

brianp committed:
Update tari crypto

brianp committed:
Add dh shared secret function

brianp committed:
Fixup instruction codes

After 9 instructions the ledger would not receive instruction 10. Had to
be changed to 0x10 to be recognized.

brianp committed:
Permit/Deny based on key type

brianp committed:
Use the device for additional diffie hellman calls

SWvheerden committed:
Merge branch 'development' into feat-ledger-recovery
